launch/launch/descriptions/executable.py

```diff
@@ -0,0 +1,141 @@
+from typing import Tuple
+
+from launch.some_substitutions_type import SomeSubstitutionsType
+from launch.substitution import Substitution
+from launch.launch_context import LaunchContext
+from launch.utilities import normalize_to_list_of_substitutions
+from launch.utilities import perform_substitutions
+
+_global_process_counter_lock = threading.Lock()
+_global_process_counter = 0  # in Python3, this number is unbounded (no rollover)
+
+class Executable:
+    """Describes an executable (typically a single process) which may be run by the launch system."""
+
+    def __init__(
+        self, *,
+        cmd: Iterable[SomeSubstitutionsType],
+        name: Optional[SomeSubstitutionsType] = None,
+        cwd: Optional[SomeSubstitutionsType] = None,
+        env: Optional[Dict[SomeSubstitutionsType, SomeSubstitutionsType]] = None,
+        additional_env: Optional[Dict[SomeSubstitutionsType, SomeSubstitutionsType]] = None,
+    ) -> None:
+        """
+        Initialize an Executable description.
+
+        :param cmd: A list where the first item is the executable and the rest are
+            arguments to the executable, each item may be a string or a list of strings
+            and Substitutions to be resolved at runtime
+        :param name: The label used to represent the process, as a string or a Substitution
+            to be resolved at runtime, defaults to the basename of the executable
+        :param cwd: The directory in which to run the executable
+        :param env: Dictionary of environment variables to be used, starting from a clean
+            environment. If None, the current environment is used.
+        :param additional_env: Dictionary of environment variables to be added. If env was
+            None, they are added to the current environment. If not, env is updated with
+            additional_env.
+        """
+        self.__cmd = [normalize_to_list_of_substitutions(x) for x in cmd]
+        self.__name = name if name is None else normalize_to_list_of_substitutions(name)
+        self.__cwd = cwd if cwd is None else normalize_to_list_of_substitutions(cwd)
+        self.__env = None  # type: Optional[List[Tuple[List[Substitution], List[Substitution]]]]
+        if env is not None:
+            self.__env = []
+            for key, value in env.items():
+                self.__env.append((
+                    normalize_to_list_of_substitutions(key),
+                    normalize_to_list_of_substitutions(value)))
+        self.__additional_env: Optional[List[Tuple[List[Substitution], List[Substitution]]]] = None
+        if additional_env is not None:
+            self.__additional_env = []
+            for key, value in additional_env.items():
+                self.__additional_env.append((
+                    normalize_to_list_of_substitutions(key),
+                    normalize_to_list_of_substitutions(value)))
+
+    @property
+    def name(self):
+        """Getter for name."""
+        return self.__name
```

Comment:
In an unrelated (or maybe not quite unrelated) note, `launch` and `launch_ros` actions are inconsistent w.r.t how action properties interact with the action lifecycle.

- `launch.actions.ExecuteProcess.name` gets you a normalized substitution, always.
- `launch_ros.actions.Node.node_name` raises if you try to get it before executing the action.

`launch_ros.parameter_description.Parameter.name`, though arguably not an action, introduces yet another variation and gets you a normalized substitution before evaluating the parameter and the actual name after evaluating it. 

@ivanpauno @wjwwood perhaps we should settle this here and stick to it.

Reply:
@hidmic Would it be worth introducing more explicitly named parameters instead, to clarify which thing you're requesting? Maybe `name` always returns the value as it was originally specified, without applying substitutions, and `applied_name` returns the one with substitutions applied to it, raising an exception if substitutions haven't been performed yet?

Reply:
> @hidmic Would it be worth introducing more explicitly named parameters instead, to clarify which thing you're requesting? Maybe name always returns the value as it was originally specified, without applying substitutions, and applied_name returns the one with substitutions applied to it, raising an exception if substitutions haven't been performed yet?

I think that's ok.
bikeshed: I'm not convinced about naming, maybe `final_name` sounds better than `applied_name`. I don't mind much.

Reply:
I've taken a stab at this in and updated. In reviewing how the objects were used, I don't think this was the right place to be creating the packed dictionary anyway, so I've removed that; it'll probably come back in `ExecuteLocal`. For all the parameters which can have substitutions applied, there's now a "`final_`" version which can be accessed. Right now, that isn't throwing an error if the substitutions haven't been applied, but I can make it do so if that's the standard approach.

Reply:
It's OK for now. It's not up to you to solve these discrepancies anyways.